README.txt
Armadillo C++ Linear Algebra Library
http://arma.sourceforge.net



Contents
========

 1: Introduction
 2: Citation Details

 3: Licenses
 4: Technical Support

 5: Requirements

 6: Linux and Mac OS X: Installation
 7: Linux and Mac OS X: Compiling & Linking

 8: Windows: Installation
 9: Windows: Compiling & Linking

10: Support for OpenBLAS, Intel MKL and AMD ACML
11: Support for ATLAS

12: API Documentation
13: MEX Interface to Octave/Matlab

14: Bug Reports and Frequently Asked Questions
15: Related Software



1: Introduction
===============


Armadillo is a high quality C++ linear algebra library,
aiming towards a good balance between speed and ease of use.

It's useful for algorithm development directly in C++,
and/or quick conversion of research code into production environments.
The syntax (API) is deliberately similar to Matlab.

The library provides efficient classes for vectors, matrices and cubes,
as well as 150+ associated functions (eg. contiguous and non-contiguous
submatrix views).  Various matrix decompositions are provided through
integration with LAPACK, or one of its high performance drop-in replacements
(eg. OpenBLAS, Intel MKL, AMD ACML, Apple Accelerate framework, etc).

An automatic expression evaluator (via C++ template meta-programming)
combines several operations (at compile time) to increase efficiency.

The library can be used for machine learning, pattern recognition,
signal processing, bioinformatics, statistics, econometrics, etc.

Armadillo is primarily developed at Data61 / NICTA (Australia).
For information about Data61 see http://data61.csiro.au

Main developers:
  Conrad Sanderson - http://conradsanderson.id.au
  Ryan Curtin      - http://ratml.org



2: Citation Details
===================

Please cite the following tech report if you use Armadillo in your
research and/or software. Citations are useful for the continued
development and maintenance of the library.

  Conrad Sanderson.
  Armadillo: An Open Source C++ Linear Algebra Library for
  Fast Prototyping and Computationally Intensive Experiments.
  Technical Report, NICTA, 2010.



3: Licenses
===========

Armadillo is available under 2 licenses:

- Open source, using the Mozilla Public License (MPL) 2.0.
  See the "LICENSE.txt" file for details.
  
- Non-open source (commercial) license, available for purchase.
  Please contact Conrad Sanderson for more information:
  http://conradsanderson.id.au



4: Technical Support
====================

You can purchase technical support on a commercial basis.
Please contact Conrad Sanderson for more information:
http://conradsanderson.id.au



5: Requirements
===============

Armadillo makes extensive use of template meta-programming, recursive templates
and template based function overloading.  As such, C++ compilers which do not
fully implement the C++ standard may not work correctly.

The functionality of Armadillo is partly dependent on other libraries:
LAPACK, BLAS, ARPACK and SuperLU.  The LAPACK and BLAS libraries are
used for dense matrices, while the ARPACK and SuperLU libraries are
used for sparse matrices.  Armadillo can work without these libraries,
but its functionality will be reduced. In particular, basic functionality
will be available (eg. matrix addition and multiplication), but things
like eigen decomposition or matrix inversion will not be.
Matrix multiplication (mainly for big matrices) may not be as fast.

As Armadillo is a template library, we recommended that optimisation
is enabled during compilation of programs that use Armadillo.
For example, for GCC and Clang compilers use -O2 or -O3



6: Linux and Mac OS X: Installation
===================================

* Step 1:
  Ensure a C++ compiler is installed on your system.
  
  Caveat: on Mac OS X you will need to install Xcode
  and then type the following command in a terminal window:
  xcode-select --install
  
* Step 2:
  Ensure the CMake tool is installed on your system.
  You can download it from http://www.cmake.org
  or (preferably) install it using your package manager.
  
  On Linux-based systems, you can get CMake using yum, dnf, apt, aptitude, ...
  
  On Mac OS X systems, you can get CMake through MacPorts or Homebrew.
  
* Step 3:
  Ensure LAPACK and BLAS are installed on your system.
  On Mac OS X this is not necessary.
  
  For better performance, we recommend installing the OpenBLAS library.
  See http://xianyi.github.com/OpenBLAS/
  
  If you are using sparse matrices, also install ARPACK and SuperLU.
  Caveat: only SuperLU version 4.3 can be used!
  
  On Linux-based systems, the following libraries are recommended
  to be present: OpenBLAS, LAPACK, SuperLU and ARPACK.
  It is also necessary to install the corresponding development
  files for each library.  For example, when installing the "lapack"
  package, also install the "lapack-devel" or "lapack-dev" package.
  
* Step 4:
  Open a terminal window, change into the directory that was created
  by unpacking the armadillo archive, and type the following commands:
  
  cmake .
  make 
  
  The full stop separated from "cmake" by a space is important.
  CMake will detect which relevant libraries are installed on your system
  (eg. OpenBLAS, LAPACK, SuperLU, ARPACK, etc)
  and will modify Armadillo's configuration correspondingly.
  CMake will also generate a run-time armadillo library,
  which is a wrapper for all the detected libraries.
  
  If you need to re-run cmake, it's a good idea to first delete the
  "CMakeCache.txt" file (not "CMakeLists.txt").
  
  Caveat: out-of-tree builds are currently not fully supported;
  eg, creating a sub-directory called "build" and running cmake ..
  from within "build" is currently not supported.
  
* Step 5:
  If you have access to root/administrator/superuser privileges
  (ie. able to use "sudo"), type the following command:
  
  sudo make install
  
  If you don't have root/administrator/superuser privileges, 
  type the following command:
  
  make install DESTDIR=my_usr_dir
  
  where "my_usr_dir" is for storing C++ headers and library files.
  Caveat: make sure your C++ compiler is configured to use the
  "lib" and "include" sub-directories present within this directory.



7: Linux and Mac OS X: Compiling & Linking
==========================================

The "examples" directory contains several quick example programs
that use the Armadillo library.

In general, programs which use Armadillo are compiled along these lines:
  
  g++ example1.cpp -o example1 -O2 -larmadillo
  
If you want to use Armadillo without installation (not recommended),
compile along these lines:
  
  g++ example1.cpp -o example1 -O2 -I /home/blah/armadillo-6.700.3/include -DARMA_DONT_USE_WRAPPER -lblas -llapack
  
The above command line assumes that you have unpacked the armadillo archive into /home/blah/
You will need to adjust this for later versions of Armadillo,
and/or if you have unpacked the armadillo archive into a different directory.
Replace -lblas with -lopenblas if you have OpenBLAS.
On Mac OS X, replace -lblas -llapack with -framework Accelerate



8: Windows: Installation
========================

The installation is comprised of 3 steps:

* Step 1:
  Copy the entire "include" folder to a convenient location
  and tell your compiler to use that location for header files
  (in addition to the locations it uses already).
  Alternatively, you can use the "include" folder directly.
  
* Step 2:
  Modify "include/armadillo_bits/config.hpp" to indicate which
  libraries are currently available on your system. For example,
  if you have LAPACK, BLAS (or OpenBLAS), ARPACK and SuperLU present,
  uncomment the following lines:
  
  #define ARMA_USE_LAPACK
  #define ARMA_USE_BLAS
  #define ARMA_USE_ARPACK
  #define ARMA_USE_SUPERLU
  
  If you don't need sparse matrices, don't worry about ARPACK or SuperLU.
  
* Step 3:
  Configure your compiler to link with LAPACK and BLAS
  (and optionally ARPACK and SuperLU).



9: Windows: Compiling & Linking
===============================

Within the "examples" folder, we have included an MSVC project named "example1_win64"
which can be used to compile "example1.cpp".  The project needs to be compiled as a
64 bit program: the active solution platform must be set to x64, instead of win32.

The MSCV project was tested on 64 bit Windows 7 with Visual C++ 2012.
You may need to make adaptations for 32 bit systems, later versions of Windows
and/or the compiler.  For example, you may have to enable or disable
ARMA_BLAS_LONG and ARMA_BLAS_UNDERSCORE macros in "armadillo_bits/config.hpp".

The folder "examples/lib_win64" contains standard LAPACK and BLAS libraries compiled
for 64 bit Windows.  The compilation was done by a third party.  USE AT YOUR OWN RISK.
The compiled versions of LAPACK and BLAS were obtained from:
  http://ylzhao.blogspot.com.au/2013/10/blas-lapack-precompiled-binaries-for.html

You can find the original sources for standard BLAS and LAPACK at:
  http://www.netlib.org/blas/
  http://www.netlib.org/lapack/
  
Faster and/or alternative implementations of BLAS and LAPACK are available:
  http://xianyi.github.com/OpenBLAS/
  http://software.intel.com/en-us/intel-mkl/
  http://developer.amd.com/tools-and-sdks/cpu-development/amd-core-math-library-acml/
  http://icl.cs.utk.edu/lapack-for-windows/lapack/

The OpenBLAS, MKL and ACML libraries are generally the fastest.

For better performance, we recommend the following high-quality C++ compilers:
  GCC from MinGW:     http://www.mingw.org/
  GCC from CygWin:    http://www.cygwin.com/
  Intel C++ compiler: http://software.intel.com/en-us/intel-compilers/

For the GCC compiler, use version 4.2 or later.
For the Intel compiler, use version 11.0 or later.

For best results we also recommend using an operating system
that's more reliable and more suitable for heavy duty work,
such as Mac OS X, or various Linux-based systems:
  Ubuntu                    http://www.ubuntu.com/
  Debian                    http://www.debian.org/
  OpenSUSE                  http://www.opensuse.org/
  Fedora                    http://fedoraproject.org/
  Scientific Linux          http://www.scientificlinux.org/
  CentOS                    http://centos.org/
  Red Hat Enterprise Linux  http://www.redhat.com/



10: Support for OpenBLAS, Intel MKL and AMD ACML
================================================

Armadillo can use OpenBLAS, or Intel Math Kernel Library (MKL),
or the AMD Core Math Library (ACML) as high-speed replacements
for BLAS and LAPACK.  Generally this just involves linking with
the replacement libraries instead of BLAS and LAPACK.

You may need to make minor modifications to "include/armadillo_bits/config.hpp"
in order to make sure Armadillo uses the same style of function names
as used by MKL or ACML. For example, the function names might be in capitals.

On Linux systems, MKL and ACML might be installed in a non-standard
location, such as /opt, which can cause problems during linking.
Before installing Armadillo, the system should know where the MKL or ACML
libraries are located. For example, "/opt/intel/mkl/lib/intel64/".
This can be achieved by setting the LD_LIBRARY_PATH environment variable,
or for a more permanent solution, adding the directory locations
to "/etc/ld.so.conf". It may also be possible to store a text file 
with the locations in the "/etc/ld.so.conf.d" directory.
For example, "/etc/ld.so.conf.d/mkl.conf".
If you modify "/etc/ld.so.conf" or create "/etc/ld.so.conf.d/mkl.conf",
you will need to run "/sbin/ldconfig" afterwards.

Example of the contents of "/etc/ld.so.conf.d/mkl.conf" on a RHEL 6 system,
where Intel MKL version 11.0.3 is installed in "/opt/intel":

/opt/intel/lib/intel64
/opt/intel/mkl/lib/intel64

The default installations of ACML 4.4.0 and MKL 10.2.2.025 are known 
to have issues with SELinux, which is turned on by default in Fedora
(and possibly RHEL). The problem may manifest itself during run-time,
where the run-time linker reports permission problems.
It is possible to work around the problem by applying an appropriate
SELinux type to all ACML and MKL libraries.

If you have ACML or MKL installed and they are persistently giving
you problems during linking, you can disable the support for them
by editing the "CMakeLists.txt" file, deleting "CMakeCache.txt" and
re-running the CMake based installation. Specifically, comment out
the lines containing:
  INCLUDE(ARMA_FindMKL)
  INCLUDE(ARMA_FindACMLMP)
  INCLUDE(ARMA_FindACML)



11: Support for ATLAS
=====================

Armadillo can use the ATLAS library for faster versions of
certain LAPACK and BLAS functions. Not all ATLAS functions are
currently used, and as such LAPACK should still be installed.

The minimum recommended version of ATLAS is 3.8.
Old versions (eg. 3.6) can produce incorrect results
as well as corrupting memory, leading to random crashes.

Users of older Ubuntu and Debian based systems should explicitly
check that ATLAS 3.6 is not installed.  It's better to
remove the old version and use the standard LAPACK library.



12: API Documentation
=====================

Documentation of functions and classes is available at:
  
  http://arma.sourceforge.net/docs.html

The documentation is also in the "docs.html" file in this archive,
which can be viewed with a web browser.



13: MEX Interface to Octave/Matlab
==================================

The "mex_interface" folder contains examples of how to interface
Octave/Matlab with C++ code that uses Armadillo matrices.



14: Bug Reports and Frequently Asked Questions
==============================================

Answers to frequently asked questions can be found at:

  http://arma.sourceforge.net/faq.html

This library has gone through extensive testing and
has been successfully used in production environments.
However, as with almost all software, it's impossible
to guarantee 100% correct functionality.

If you find a bug in the library (or the documentation),
we are interested in hearing about it. Please make a
_small_ and _self-contained_ program which exposes the bug,
and then send the program source (as well as the bug description)
to the developers.  The developers' contact details are at:

  http://arma.sourceforge.net/contact.html



15: Related Software
====================

* MLPACK: C++ library for machine learning and pattern recognition, built on top of Armadillo.
  http://mlpack.org
  
* Mantella: C++ library for analysing and solving optimisation problems
  https://github.com/SebastianNiemann/Mantella
  
* libpca: C++ library for principal component analysis
  http://sourceforge.net/projects/libpca/
  
* ArmaNpy: interfaces Armadillo matrices with Python
  http://sourceforge.net/projects/armanpy/
  
* matlab2cpp: conversion of Matlab code to Armadillo based C++ code
  https://github.com/jonathf/matlab2cpp


debian/changelog
armadillo (1:10.8.2+dfsg-1) unstable; urgency=medium

  * New upstream release

 -- Kumar Appaiah <akumar@debian.org>  Mon, 07 Feb 2022 21:02:29 +0530

armadillo (1:10.8.0+dfsg-1) unstable; urgency=medium

  * New upstream release

 -- Kumar Appaiah <akumar@debian.org>  Wed, 12 Jan 2022 20:59:36 +0530

armadillo (1:10.7.4+dfsg-1) unstable; urgency=medium

  * New upstream release

 -- Kumar Appaiah <akumar@debian.org>  Thu, 25 Nov 2021 22:28:58 +0530

armadillo (1:10.6.2+dfsg-1) unstable; urgency=medium

  * New upstream release
  * Remove all patches (included upstream)

 -- Kumar Appaiah <akumar@debian.org>  Tue, 24 Aug 2021 19:47:25 +0530

armadillo (1:10.1.2+dfsg-6) unstable; urgency=medium

  * Add patch from upstream for fixing sparse matrix related bug.

 -- Kumar Appaiah <akumar@debian.org>  Mon, 21 Jun 2021 09:43:00 +0530

armadillo (1:10.1.2+dfsg-5) unstable; urgency=medium

  * Add patches from upstream for pinv and rcond bugs

 -- Kumar Appaiah <akumar@debian.org>  Tue, 09 Mar 2021 20:19:43 +0530

armadillo (1:10.1.2+dfsg-4) unstable; urgency=medium

  * Fix issue with princomp function.

 -- Kumar Appaiah <akumar@debian.org>  Mon, 15 Feb 2021 14:52:17 +0530

armadillo (1:10.1.2+dfsg-3) unstable; urgency=medium

  * Add patch to handle sign for NaNs

 -- Kumar Appaiah <akumar@debian.org>  Wed, 20 Jan 2021 16:19:03 +0530

armadillo (1:10.1.2+dfsg-2) unstable; urgency=medium

  * Install cmake and pkg-config files during install. (Closes: #954927)
  * Remove generated files during cleanup
  * Standards version is now 4.5.1 (no changes needed)

 -- Kumar Appaiah <akumar@debian.org>  Thu, 17 Dec 2020 07:58:31 +0530

armadillo (1:10.1.2+dfsg-1) unstable; urgency=medium

  * New upstream release

 -- Kumar Appaiah <akumar@debian.org>  Wed, 18 Nov 2020 14:34:07 +0530

armadillo (1:10.1.0+dfsg-1) experimental; urgency=medium

  * New upstream release
  * Standards version is now 4.5.0

 -- Kumar Appaiah <akumar@debian.org>  Fri, 09 Oct 2020 14:59:12 +0530

armadillo (1:9.900.1+dfsg-1) unstable; urgency=medium

  * New upstream release

 -- Kumar Appaiah <akumar@debian.org>  Thu, 11 Jun 2020 14:27:04 +0530

armadillo (1:9.800.4+dfsg-1) unstable; urgency=medium

  * New upstream release

 -- Kumar Appaiah <akumar@debian.org>  Wed, 22 Jan 2020 20:29:53 +0530

armadillo (1:9.800.1+dfsg-1) unstable; urgency=medium

  * New upstream release
  * Standards version is now 4.4.0 (no changes needed)

 -- Kumar Appaiah <akumar@debian.org>  Wed, 09 Oct 2019 18:57:37 +0530

armadillo (1:9.600.5+dfsg-1) unstable; urgency=medium

  * New upstream release

 -- Kumar Appaiah <akumar@debian.org>  Mon, 29 Jul 2019 08:58:55 +0530

armadillo (1:9.200.7+dfsg-1) unstable; urgency=medium

  * New upstream release (Closes: #920015)

 -- Kumar Appaiah <akumar@debian.org>  Mon, 21 Jan 2019 21:22:55 +0530

armadillo (1:9.200.6+dfsg-1) unstable; urgency=medium

  * New upstream release

 -- Kumar Appaiah <akumar@debian.org>  Tue, 11 Dec 2018 10:47:39 +0530

armadillo (1:9.100.6+dfsg-1) unstable; urgency=medium

  * New upstream release
  * Update watch file to avoid RC releases.

 -- Kumar Appaiah <akumar@debian.org>  Mon, 29 Oct 2018 22:04:27 +0530

armadillo (1:9.100.5+dfsg-1) unstable; urgency=medium

  * New upstream release
  * soname is now 9, add libarmadillo9 package
  * Remove band mattrix patch

 -- Kumar Appaiah <akumar@debian.org>  Thu, 16 Aug 2018 14:52:34 +0530

armadillo (1:8.400.0+dfsg-3) unstable; urgency=medium

  * Add patch from upstream to fix band matrix bug.
  * Move VCS details to Debian Salsa.

 -- Kumar Appaiah <akumar@debian.org>  Fri, 18 May 2018 16:54:54 +0530

armadillo (1:8.400.0+dfsg-2) unstable; urgency=medium

  * Fix debian/rules to allow armadillo to build from source. Patch from
    Helmut Grohne. (Closes: #875405)
  * Add -V to dh_makeshlibs to obtain versioned dependencies. (Closes: #885971)
  * debhelper compay 9 automagically produces debug symbols.. (Closes: #744048)

 -- Kumar Appaiah <akumar@debian.org>  Wed, 21 Feb 2018 06:04:02 +0530

armadillo (1:8.400.0+dfsg-1) unstable; urgency=medium

  * New upstream release

 -- Kumar Appaiah <akumar@debian.org>  Tue, 20 Feb 2018 19:48:07 +0530

armadillo (1:8.300.3+dfsg-1) unstable; urgency=medium

  * New upstream release

 -- Kumar Appaiah <akumar@debian.org>  Mon, 22 Jan 2018 03:40:01 +0530

armadillo (1:8.300.1+dfsg-1) unstable; urgency=medium

  * New upstream release

 -- Kumar Appaiah <akumar@debian.org>  Sun, 10 Dec 2017 09:43:58 +0530

armadillo (1:8.200.2+dfsg-1) unstable; urgency=medium

  * New upstream release

 -- Kumar Appaiah <akumar@debian.org>  Tue, 21 Nov 2017 07:14:33 +0530

armadillo (1:8.200.0+dfsg-3) unstable; urgency=medium

  * Upload to unstable
  * Fix copyright file to adhere to copyright format 1.0

 -- Kumar Appaiah <akumar@debian.org>  Mon, 06 Nov 2017 11:59:34 +0530

armadillo (1:8.200.0+dfsg-2) experimental; urgency=medium

  * Update and fix copyright to reflect new and corrected license and
    copyright holder details.

 -- Kumar Appaiah <akumar@debian.org>  Wed, 11 Oct 2017 08:56:05 +0530

armadillo (1:8.200.0+dfsg-1) experimental; urgency=medium

  * New upstream release
  * The new package (with soname) is libarmadillo8
  * Standards Version is now 4.1.1 (no changes)

 -- Kumar Appaiah <akumar@debian.org>  Tue, 10 Oct 2017 05:22:15 +0530

armadillo (1:7.960.1+dfsg-1) unstable; urgency=medium

  * New upstream release

 -- Kumar Appaiah <akumar@debian.org>  Wed, 23 Aug 2017 22:38:28 +0530

armadillo (1:7.950.1+dfsg-1) unstable; urgency=medium

  * New upstream release

 -- Kumar Appaiah <akumar@debian.org>  Tue, 20 Jun 2017 20:36:34 +0530

armadillo (1:7.900.1+dfsg-2) experimental; urgency=medium

  * Update copyright to reflect Apache License

 -- Kumar Appaiah <akumar@debian.org>  Tue, 16 May 2017 07:08:51 +0530

armadillo (1:7.900.1+dfsg-1) experimental; urgency=medium

  * New upstream release

 -- Kumar Appaiah <akumar@debian.org>  Mon, 15 May 2017 20:42:34 +0530


include/armadillo_bits/config.hpp
// Copyright (C) 2008-2012 NICTA (www.nicta.com.au)
// Copyright (C) 2008-2012 Conrad Sanderson
// 
// This file is part of the Armadillo C++ library.
// It is provided without any warranty of fitness
// for any purpose. You can redistribute this file
// and/or modify it under the terms of the GNU
// Lesser General Public License (LGPL) as published
// by the Free Software Foundation, either version 3
// of the License or (at your option) any later version.
// (see http://www.opensource.org/licenses for more info)



#if !defined(ARMA_USE_LAPACK)
// #define ARMA_USE_LAPACK
//// Uncomment the above line if you have LAPACK or a high-speed replacement for LAPACK,
//// such as Intel's MKL, AMD's ACML, or the Accelerate framework.
//// LAPACK is required for matrix decompositions (eg. SVD) and matrix inverse.
#endif

#if !defined(ARMA_USE_BLAS)
// #define ARMA_USE_BLAS
//// Uncomment the above line if you have BLAS or a high-speed replacement for BLAS,
//// such as GotoBLAS, Intel's MKL, AMD's ACML, or the Accelerate framework.
//// BLAS is used for matrix multiplication.
//// Without BLAS, matrix multiplication will still work, but might be slower.
#endif

// #define ARMA_BLAS_CAPITALS
//// Uncomment the above line if your BLAS and LAPACK libraries have capitalised function names (eg. ACML on 64-bit Windows)

#define ARMA_BLAS_UNDERSCORE
//// Uncomment the above line if your BLAS and LAPACK libraries have function names with a trailing underscore.
//// Conversely, comment it out if the function names don't have a trailing underscore.

// #define ARMA_BLAS_LONG
//// Uncomment the above line if your BLAS and LAPACK libraries use "long" instead of "int"

// #define ARMA_BLAS_LONG_LONG
//// Uncomment the above line if your BLAS and LAPACK libraries use "long long" instead of "int"

// #define ARMA_USE_TBB_ALLOC
//// Uncomment the above line if you want to use Intel TBB scalable_malloc() and scalable_free() instead of standard new[] and delete[]

// #define ARMA_USE_MKL_ALLOC
//// Uncomment the above line if you want to use Intel MKL mkl_malloc() and mkl_free() instead of standard new[] and delete[]

// #define ARMA_USE_ATLAS
// #define ARMA_ATLAS_INCLUDE_DIR /usr/include/
//// If you're using ATLAS and the compiler can't find cblas.h and/or clapack.h
//// uncomment the above define and specify the appropriate include directory.
//// Make sure the directory has a trailing /

// #define ARMA_64BIT_WORD
//// Uncomment the above line if you require matrices/vectors capable of holding more than 4 billion elements.
//// Your machine and compiler must have support for 64 bit integers (eg. via "long" or "long long")

// #define ARMA_USE_CXX11
//// Uncomment the above line if you have a C++ compiler that supports the C++11 standard
//// This will enable additional features, such as use of initialiser lists

#if !defined(ARMA_MAT_PREALLOC)
  #define ARMA_MAT_PREALLOC 16
#endif
//// This is the number of preallocated elements used by matrices and vectors;
//// it must be an integer that is at least 1.
//// If you mainly use lots of very small vectors (eg. <= 4 elements),
//// change the number to the size of your vectors.

#if !defined(ARMA_SPMAT_CHUNKSIZE)
  #define ARMA_SPMAT_CHUNKSIZE 256
#endif
//// This is the minimum increase in the amount of memory (in terms of elements) allocated by a sparse matrix;
//// it must be an integer that is at least 1.
//// The minimum recommended size is 16.

// #define ARMA_NO_DEBUG
//// Uncomment the above line if you want to disable all run-time checks.
//// This will result in faster code, but you first need to make sure that your code runs correctly!
//// We strongly recommend to have the run-time checks enabled during development,
//// as this greatly aids in finding mistakes in your code, and hence speeds up development.
//// We recommend that run-time checks be disabled _only_ for the shipped version of your program.

// #define ARMA_EXTRA_DEBUG
//// Uncomment the above line if you want to see the function traces of how Armadillo evaluates expressions.
//// This is mainly useful for debugging of the library.

// #define ARMA_USE_BOOST
// #define ARMA_USE_BOOST_DATE
// #define ARMA_USE_WRAPPER
// #define ARMA_USE_HDF5

#if !defined(ARMA_DEFAULT_OSTREAM)
  #define ARMA_DEFAULT_OSTREAM std::cout
#endif

#define ARMA_PRINT_LOGIC_ERRORS
#define ARMA_PRINT_RUNTIME_ERRORS
//#define ARMA_PRINT_HDF5_ERRORS

// #define ARMA_HAVE_STD_ISFINITE
// #define ARMA_HAVE_STD_ISINF
// #define ARMA_HAVE_STD_ISNAN
// #define ARMA_HAVE_STD_SNPRINTF

#define ARMA_HAVE_LOG1P
#define ARMA_HAVE_GETTIMEOFDAY



#if defined(ARMA_DONT_USE_LAPACK)
  #undef ARMA_USE_LAPACK
#endif

#if defined(ARMA_DONT_USE_BLAS)
  #undef ARMA_USE_BLAS
#endif

#if defined(ARMA_DONT_USE_ATLAS)
  #undef ARMA_USE_ATLAS
  #undef ARMA_ATLAS_INCLUDE_DIR
#endif

#if defined(ARMA_DONT_PRINT_LOGIC_ERRORS)
  #undef ARMA_PRINT_LOGIC_ERRORS
#endif

#if defined(ARMA_DONT_PRINT_RUNTIME_ERRORS)
  #undef ARMA_PRINT_RUNTIME_ERRORS
#endif
